### PR TITLE
Code fix for stat issues when sched starts first in Mulit-Server environment

### PR DESCRIPTION
**Pull request by suresh-thelkar**
#### Describe Bug or Feature

This PR addresses the following.

- Some of the stats failing from sched when it is started first before Server in Multi-Server environment
- Fix for race conditions in Scheduler's secondary connections.
- This PR handles race conditions when lot of failures happen at the Scheduler which could be because of stats failing from scheduler or any other reason. 

#### Attach Test and Valgrind Logs/Output

Attached are test logs.
[PR108.logs.txt](https://github.com/subhasisb/openpbs/files/4930214/PR108.logs.txt)

### CODE REVIEW

File: src/scheduler/pbs_sched.c [hunk 1]
```diff
@@ -1595,7 +1599,6 @@ schedule_wrapper(int num_cfg_svrs, int *update_svr,fd_set *read_fdset, int opt_n
 				if (update_svr != NULL && (*update_svr)) {
 					/* update sched object attributes on server */
 					if (update_svr_schedobj(svr_conns[0]->sd, cmd, alarm_time) == 0) {
-						send_cycle_end(second_connection);
```
Comment (agrawalravi90) [inline src/scheduler/pbs_sched.c:1598]: why did you remove the calls to send_cycle_end?

Comment (suresh-thelkar) [reply src/scheduler/pbs_sched.c:1598]: Whenever we are calling close_server_conn() we actually no need to send end of cycle notification to the respective server. It is because closing server connection itself acts as a notification to the server after which server resets psched->sched_cycle_started back to 0 so that it can send further scheduling commands when sched is ready.

**Comment by suresh-thelkar** [comment]
PR is ready for further reviews.

**Comment by suresh-thelkar** [comment]
@subhasisb, Can you please take a look and merge this PR if it looks ok.

File: src/scheduler/pbs_sched.c [hunk 2]
```diff
@@ -322,15 +322,16 @@ close_server_conn(int index_to_shards)
...
-			close_tcp_connection(svr_conns[index_to_shards]->secondary_sd);
-			/* unlock the connection level lock */
-			svr_conns[index_to_shards]->secondary_sd = -1;
-		}
 		if (svr_conns[index_to_shards]->sd >= 0) {
 			close_tcp_connection(svr_conns[index_to_shards]->sd);
 			/* unlock the connection level lock */
 			svr_conns[index_to_shards]->sd = -1;
+			if (svr_conns[index_to_shards]->secondary_sd >= 0) {
+				CS_close_socket(svr_conns[index_to_shards]->secondary_sd);
```
Comment (agrawalravi90) [inline src/scheduler/pbs_sched.c:330]: why not use close_tcp_connection() like before?

Comment (suresh-thelkar) [reply src/scheduler/pbs_sched.c:329]: There is no need to use close_tcp_connection() on secondary connection also. It is because when primary connection is closed we send PBS_BATCH_Disconnect to Server, upon receiving this message server does the necessary clean up and closes the connection at its side. So we no need to call PBS_BATCH_Disconnect once again on secondary connection, just closing its sd is good enough. The other reason is Server is expecting only end of cycle indication on Secondary and it we send this also the receiving sockets buffer might get corrupted or the data in it is garbled.

**Comment by suresh-thelkar** [comment]
@agrawalravi90, I have added few more changes apart from what we discussed. Can you please take a look at the latest commit and provide your review comments.

File: src/scheduler/pbs_sched.c (hunk 2 above)
Comment (agrawalravi90) [reply src/scheduler/pbs_sched.c:329]: can you please add a comment here explaining briefly that we are intentionally not calling close_tcp_connection() here?

File: src/scheduler/fifo.c [hunk 3]
```diff
@@ -3013,7 +3013,7 @@ set_validate_sched_attrs(int connector)
 
 	/* Stat the scheduler to get details of sched */
 
-	all_ss = pbs_statsched(connector, NULL, NULL);
+	all_ss = pbs_statsched(entry_to_svr_conns, NULL, NULL);
```
[the remark this answers is not in the document]
Comment (suresh-thelkar) [reply src/scheduler/fifo.c:3016]: pbs_statsched uses get_conn_servers() to access the svr_conns table for which we need entry_to_svr_conns. 
If we use connector instead which will give problem if we have more than one servers. Let's say we have two servers and when connector belongs to second server for an iteration then when we call get_conn_servers() which returns NULL.

File: src/scheduler/globals.h [hunk 4]
```diff
@@ -108,6 +108,8 @@ extern time_t last_attr_updates;    /* timestamp of the last time attr updates w
 
 extern int send_job_attr_updates;
 
+extern int entry_to_svr_conns;
```
[the remark this answers is not in the document]
Comment (suresh-thelkar) [reply src/scheduler/globals.h:111]: This variable is global because it is used inside update_svr_schedobj() for which we only pass actual socket i.e. sock_to_check instead of entry_to_svr_conns. And also this variable is needed in fifo.c i.e. wherever set_validate_sched_attrs is called.
I observed that you are eliminating entry_to_svr_conns in your on going PR in which case it is not a problem. Let me know what do you think.

File: src/scheduler/pbs_sched.c (hunk 2 above)
Comment (suresh-thelkar) [reply src/scheduler/pbs_sched.c:333]: I have added a detail comment.

**Comment by suresh-thelkar** [comment]
@agrawalravi90, I have addressed your comments. Please take a look.

File: src/scheduler/fifo.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (agrawalravi90) [reply src/scheduler/fifo.c:3016]: when you call update_svr_schedobj() from schedule_wrapper(), you are passing sock_to_check as the "connector" .. if you don't use connector here, then it will just send sched port and host to the entry_to_svr_conns server, or a random server,  right?

> If we use connector instead which will give problem if we have more than one servers. Let's say we have two servers and when connector belongs to second server for an iteration then when we call get_conn_servers() which returns NULL.
Two things here:
- you are passing entry_to_svr_conns to schedule(), so "connector" will be "entry_to_svr_conns" anyways when we are running scheduling cycles. The only place where it will be specific to the server is from the call to update_svr_schedobj(), and there we need it to be different.
- After my changes, get_conn_servers() will always return the array of server connections for this thread, there's only going to be one array for each thread, so it will not fail no matter what fd you pass to it.

File: src/scheduler/globals.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (agrawalravi90) [reply src/scheduler/globals.h:111]: As I explained above, I think both update_svr_schedobj() and set_validate_sched_attrs() need the connector to be different than entry_to_svr_conns so that the pbs_manager call which updates server port and host goes only to that specific server, not to any random server, or the server associated with entry_to_svr_conns fd.

**Comment by subhasisb** [comment]
@agrawalravi90 can you look at it from functionality perspective - we can refactor the code to be nicer later - for now, i just need things to be working - thats it

**Comment by agrawalravi90** [comment]
@subhasisb there are some important comments left to be resolved, once they are resolved ill approve it

File: src/scheduler/fifo.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (suresh-thelkar) [reply src/scheduler/fifo.c:3016]: > when you call update_svr_schedobj() from schedule_wrapper(), you are passing sock_to_check as the "connector" .. if you don't use connector here, then it will just send sched port and host to the entry_to_svr_conns server, or a random server, right?

I am actually making use of connector also so it will send port snd host to the right server. As we are syncing up today we can discuss more on this and come to a common page.

File: src/scheduler/globals.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (suresh-thelkar) [reply src/scheduler/globals.h:111]: update_svr_schedobj() is actually sending the data to the right server because I am making use of connector as well as entry_to_svr_conns whichever is appropriate. As we are syncing up we can discuss more on this and come to a common page.

**Comment by suresh-thelkar** [comment]
@agrawalravi90 , Thanks for your sign off.
